src/feature/hs/hs_cell.c

```diff
@@ -853,17 +853,46 @@ hs_cell_parse_introduce2(hs_cell_introduce2_data_t *data,
 
   /* First bytes of the ENCRYPTED section are the client public key (they are
    * guaranteed to exist because of the length check above). We are gonna use
-   * the client public key to compute the ntor keys and decrypt the payload: */
-  memcpy(&data->client_pk.public_key, encrypted_section, CURVE25519_PUBKEY_LEN);
-
-  /* Get the right INTRODUCE2 ntor keys and verify the cell MAC */
-  intro_keys = get_introduce2_keys_and_verify_mac(data, encrypted_section,
-                                                  encrypted_section_len);
-  if (!intro_keys) {
-    log_info(LD_REND, "Could not get valid INTRO2 keys on circuit %u "
-             "for service %s", TO_CIRCUIT(circ)->n_circ_id,
-             safe_str_client(service->onion_address));
-    goto done;
+   * the client public key to compute the ntor keys and decrypt the payload:
+   */
+  memcpy(&data->client_pk.public_key, encrypted_section,
+         CURVE25519_PUBKEY_LEN);
+
+  /* If we are configured as an Onion Balance instance, we need to try
+   * validation with the configured master public keys given in the config
+   * file. This is because the master identity key and the blinded key is put
+   * in the INTRODUCE2 cell by the client thus it will never validate with
+   * this instance public key. */
+  if (service->config.ob_master_pubkeys) {
+    hs_cell_introduce2_data_t new_data;
+
+    SMARTLIST_FOREACH_BEGIN(service->config.ob_master_pubkeys,
+                            const ed25519_public_key_t *, pkey) {
+      ed25519_public_key_t blinded_pubkey;
+      uint8_t subcredential[DIGEST256_LEN];
+
+      hs_build_blinded_pubkey(pkey, NULL, 0, hs_get_time_period_num(0),
+                              &blinded_pubkey);
+      hs_get_subcredential(pkey, &blinded_pubkey, subcredential);
+      memcpy(&new_data, data, sizeof(new_data));
```

Comment:
OK this works just fine in Chutney! However, it seems like `data->client_pk` is zeroed out and that makes `hs_ntor_service_get_rendezvous1_keys()` fail in:
```
  curve25519_handshake(dh_result1,
                       &service_ephemeral_rend_keypair->seckey,
                       client_ephemeral_enc_pubkey);
```

I'm not sure what's going on but I fixed this by commenting out the `memwipe(&data->client_pk, 0, sizeof(curve25519_public_key_t));` in `hs_cell_parse_introduce2()`. It worked after this and I managed to actually connect to onionbalance and pass data around.

Reply:
Ah the bug here seems to be introduced by the memwipe I added in the refactoring commit ee3f36476a87cab24881d56b8c9227c80af19521 . Perhaps we can kill that commit? Or fix it? Up to you.

Reply:
Agree. I've removed it for now in the rebased version of this branch.